biz.aQute.bndlib/src/aQute/bnd/osgi/PluginsContainer.java

```diff
@@ -40,24 +45,112 @@
 import aQute.libg.cryptography.SHA1;
 
 /**
- * The plugin set for a Processor.
+ * The plugin set for a Processor. Plugins are general service objects and can
+ * be any type. The PluginsContainer treats the @{link {@link PluginProvider}
+ * special. If it is used with a type, it will expand the when it encounters a
+ * {@link ProviderType} plugin, see {@link #getPlugin(Class)} and
+ * {@link #getPlugin(Class)}. These provided plugins are _not_ part of the this
+ * set so they won't be explicitly returned when {@link #plugins()} is called.
```

Comment:
Right. So anyone who looks at the container as a `Set` (iterates, spliterates, contains, size, toArray, ...) will observe the PluginProvider. For example, the bnd `plugins` command iterates over the set and wont display the information on the actual plugins (it will display the toString of AbstractPlugin). The WorkspaceRepository tooltip will do the same.

Perhaps we can do better. The internal set can be `Set<PluginProvider<Object>>` where we wrap simple plugins in a provider. Then at all set method PluginContainer, we unwrap the PluginProvider? If you want to go this way, I can make a PR with this abstaction, and then you could rebase on that?